Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp

```diff
@@ -0,0 +1,184 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "Acts/Geometry/GeometryHierarchyMap.hpp"
+#include "ActsExamples/Digitization/DigitizationConfig.hpp"
+#include "ActsExamples/EventData/Measurement.hpp"
+#include "ActsExamples/EventData/SimHit.hpp"
+#include "ActsExamples/Framework/BareAlgorithm.hpp"
+#include "ActsExamples/Framework/RandomNumbers.hpp"
+#include "ActsFatras/Digitization/Channelizer.hpp"
+#include "ActsFatras/Digitization/PlanarSurfaceDrift.hpp"
+#include "ActsFatras/Digitization/PlanarSurfaceMask.hpp"
+
+#include <memory>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <variant>
+#include <vector>
+
+namespace Acts {
+class Surface;
+class TrackingGeometry;
+}  // namespace Acts
+
+namespace ActsExamples {
+
+/// Algorithm that turns simulated hits into measurements by truth smearing.
+class DigitizationAlgorithm final : public BareAlgorithm {
+ public:
+  struct Config {
+    /// Input collection of simulated hits.
+    std::string inputSimHits;
+    /// Output source links collection.
+    std::string outputSourceLinks;
+    /// Output measurements collection.
+    std::string outputMeasurements;
+    /// Output collection to map measured hits to contributing particles.
+    std::string outputMeasurementParticlesMap;
+    /// Output collection to map measured hits to simulated hits.
+    std::string outputMeasurementSimHitsMap;
+    /// Tracking geometry required to access global-to-local transforms.
+    std::shared_ptr<const Acts::TrackingGeometry> trackingGeometry = nullptr;
+    /// Random numbers tool.
+    std::shared_ptr<const RandomNumbers> randomNumbers = nullptr;
+    /// The digitizers per GeometryIdentifier
+    Acts::GeometryHierarchyMap<DigitizationConfig> digitizationConfigs;
+  };
+
+  /// Construct the smearing algorithm.
+  ///
+  /// @param cfg is the algorithm configuration
+  /// @param lvl is the logging level
+  DigitizationAlgorithm(Config cfg, Acts::Logging::Level lvl);
+
+  /// Build measurement from simulation hits at input.
+  ///
+  /// @param ctx is the algorithm context with event information
+  /// @return a process code indication success or failure
+  ProcessCode execute(const AlgorithmContext& ctx) const final override;
+
+ private:
+  /// Nested struct for digitized parameters
+  struct DigitizedParameters {
+    std::vector<Acts::BoundIndices> indices = {};
+    std::vector<Acts::ActsScalar> values = {};
+    std::vector<Acts::ActsScalar> covariances = {};
```

Comment:
I guess this vector is for just the errors of the parameters (not their correlation), right? Personally, I would expect a matrix from the name `covariance`. Is it better to rename it to e.g. `errors` or `uncertainties`?

Reply:
Measurement directions are by all means independent for our purpose.
We could write 'variances' ?

Reply:
Sounds good!